File: obsolete/Palladio.ComponentModel.All/trunk/ReflectionBasedVisitor/ReflectiveVisitor.cs
using System;
using System.Reflection;
using ReflectionBasedVisitor;

namespace ReflectionBasedVisitor
{
	/// <summary>
	/// The base class for a reflection based visitor pattern.
	/// See <a href="http://www.javaworld.com/javaworld/javatips/jw-javatip98.html">
	/// http://www.javaworld.com/javaworld/javatips/jw-javatip98.html</a> for details 
	/// on the usage.
	/// </summary>
	/// <remarks>
	/// A usage example in a client using this assembly
	/// <code>
	/// class MyType : IVisitable
	/// {
	/// 	public virtual void AcceptVisitor(IVisitor v) 
	/// 	{
	/// 		// Add code here if you need data structure driven visiting!
	/// 		// Otherwise leave empty!
	///		}
	/// }
	/// 
///	class MyVisitor : ReflectionBasedVisitor.ReflectionBasedVisitor, IVisitor
///	{
///		public override void VisitObject(object o)
///		{
///		   // Add default behaviour here if the visitor visits an object of unknown
///		   // type
///		}
///
///		public void VisitMyType(MyType objectOfMyType)
///		{
///		   // Do sth. when visiting an obejct of type MyType
///		   // For example:
///		   Console.WriteLine("Visiting "+objectOfMyType);
///		   Visit(objectOfMyType.Next);
///		}
///
///	}
	/// </code>
	/// <pre>
	/// Version history:
	///
	/// $Log$
	/// Revision 1.1  2005/02/21 11:57:06  joemal
	/// initial import
	///
	///
	/// </pre>
	/// </remarks>
	public abstract class ReflectionBasedVisitor : IVisitor
	{
		/// <summary>
		///Determines if the visitor moves by the way
		/// determined by the data structure (true) or if it moves on its own (false).
		/// </summary>
		protected bool isDataDrivenVisitor = false;

		/// <summary>
		/// Default constructor of the base class
		/// </summary>
		/// <param name="isDataDrivenVisitor">Determines if the visitor moves by the way
		/// determined by the data structure (true) or if it moves on its own (false).</param>
		public ReflectionBasedVisitor(bool isDataDrivenVisitor)
		{
			this.isDataDrivenVisitor = isDataDrivenVisitor;
		}

		/// <summary>
		/// The default function which is called if no "VisitTYPE" method is found
		/// by the reflection API
		/// </summary>
		/// <param name="o">The object calling the unknown visitor function</param>
		public abstract void VisitObject (object o);

		/// <summary>
		/// The Visit method called by the reflection API. Do not alter this!!!
		/// </summary>
		/// <param name="o">The object beening visited</param>
		public void Visit(object o)
		{
			MethodInfo method = GetMethod(o.GetType());
			method.Invoke(this, new object[] {o});
			if (o is IVisitable && isDataDrivenVisitor)
			{
				CallAccept((IVisitable)o);
			}
		}
		
		private void CallAccept(IVisitable visitable) 
		{
			visitable.AcceptVisitor(this);
		}

		private MethodInfo GetMethod(Type type) 
		{
			MethodInfo m = null;
			string method = "Visit" + type.Name;
			
			m = GetType().GetMethod(method, new Type[] {type});
			
			// Try the interfaces.  If necessary, you
			// can sort them first to define 'visitable' interface wins
			// in case an object implements more than one.
			if (m == null) 
			{
				Type[] interfaces = type.GetInterfaces();
				for (int i = 0; i < interfaces.Length; i++) 
				{
					method = "Visit" + interfaces[i].Name;
					m = GetType().GetMethod(method, new Type[] {interfaces[i]});
					if (m != null)
						break;
				}
			}
			if (m == null) 
			{
				m = GetType().GetMethod("VisitObject", new Type[] {Type.GetType("System.Object")});
			}
			return m;
		}
	}
}

File: obsolete/Palladio.SimuCom/src/de/uka/ipd/sdq/simucom/reflectivevisitor/ReflectiveVisitor.java
package de.uka.ipd.sdq.simucom.reflectivevisitor;

import java.lang.reflect.Method;
import java.util.HashMap;

public class ReflectiveVisitor {

	private static HashMap<Class,HashMap<Class,Method>> cache = new HashMap<Class, HashMap<Class,Method>>();
		
	public void visit(Object object) throws Exception {
		Method method = getMethodWithCache(object.getClass());
		method.invoke(this, new Object[] { object });
		if (object instanceof Visitable) {
			callAccept((Visitable) object);
		}
	}

	private Method getMethodWithCache(Class typeOfObjectToVisit) {
		HashMap<Class,Method> thisClassCache = cache.get(this.getClass());
		if (thisClassCache == null)
		{
			thisClassCache = new HashMap<Class, Method>();
			cache.put(this.getClass(),thisClassCache);
		}
		if (!thisClassCache.containsKey(typeOfObjectToVisit))
		{
			thisClassCache.put(typeOfObjectToVisit,getMethod(typeOfObjectToVisit));
		}
		return thisClassCache.get(typeOfObjectToVisit);
	}

	public void visitObject(Object o) throws Exception
	{
		throw new Exception("Object "+o.toString()+" is not visitable by this visitor "+this.toString());
	}
	
	public void callAccept(Visitable visitable) {
	   visitable.accept(this);
	}	
	
	protected Method getMethod(Class c) {
		Class newc = c;
		Method m = null;
		// Try the superclasses
		while (m == null && newc != Object.class) {
			String method = newc.getName();
			method = "visit" + method.substring(method.lastIndexOf('.') + 1);
			try {
				m = getClass().getMethod(method, new Class[] { newc });
			} catch (NoSuchMethodException e) {
				newc = newc.getSuperclass();
			}
		}
		// Try the interfaces. If necessary, you
		// can sort them first to define 'visitable' interface wins
		// in case an object implements more than one.
		if (newc == Object.class) {
			Class[] interfaces = c.getInterfaces();
			for (int i = 0; i < interfaces.length; i++) {
				String method = interfaces[i].getName();
				method = "visit"
						+ method.substring(method.lastIndexOf('.') + 1);
				try {
					m = getClass().getMethod(method,
							new Class[] { interfaces[i] });
				} catch (NoSuchMethodException e) {
				}
			}
		}
		if (m == null) {
			try {
				m = this.getClass().getMethod("visitObject",
						new Class[] { Object.class });
			} catch (Exception e) {
				// Can't happen
			}
		}
		return m;
	}
}
